spark/dl/src/main/scala/com/intel/analytics/bigdl/dlframes/DLImageTransformer.scala

```diff
@@ -0,0 +1,93 @@
+import org.apache.spark.ml.adapter.{HasInputCol, HasOutputCol, SchemaUtils}
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SqlAdapter}
+
+/**
+ * Provides DataFrame-based API for image pre-processing and feature transformation.
+ * DLImageTransformer follows the Spark Transformer API pattern and can be used as one stage
+ * in Spark ML pipeline.
+ *
+ * The input column can be either DLImageSchema.byteSchema or DLImageSchema.floatSchema. If
+ * using DLImageReader, the default format is DLImageSchema.byteSchema
+ * The output column is always DLImageSchema.floatSchema.
+ *
+ * @param transformer Single or a sequence of BigDL FeatureTransformers to be used. E.g.
+ *                    Resize(256, 256) -> CenterCrop(224, 224) ->
+ *                    ChannelNormalize(123, 117, 104, 1, 1, 1) -> MatToTensor()
+ */
+class DLImageTransformer (
+    val transformer: Transformer[ImageFeature, ImageFeature],
+    override val uid: String)
+  extends DLTransformerBase with HasInputCol with HasOutputCol {
+
+  def this(transformer: FeatureTransformer) =
+    this(transformer, Identifiable.randomUID("DLImageTransformer"))
+
+  setDefault(inputCol -> "image")
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  protected def validateInputType(inputType: DataType): Unit = {
+    val validTypes = Array(DLImageSchema.floatSchema, DLImageSchema.byteSchema)
+
+    require(validTypes.exists(t => SchemaUtils.sameType(inputType, t)),
+      s"Bad input type: $inputType. Requires ${validTypes.mkString(", ")}")
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    val inputType = schema($(inputCol)).dataType
+    validateInputType(inputType)
+    if (schema.fieldNames.contains($(outputCol))) {
+      throw new IllegalArgumentException(s"Output column ${$(outputCol)} already exists.")
+    }
+
+    val outputFields = schema.fields :+
+      StructField($(outputCol), DLImageSchema.floatSchema, nullable = false)
+    StructType(outputFields)
+  }
+
+  protected override def internalTransform(dataFrame: DataFrame): DataFrame = {
+    transformSchema(dataFrame.schema, logging = true)
+    val sc = dataFrame.sqlContext.sparkContext
+    val localTransformer = this.transformer
+    val transformerBC = sc.broadcast(localTransformer)
+
+    val transFunc = { row: Row =>
+      val transformerValue = transformerBC.value
+      val imf = DLImageSchema.row2IMF(row)
+      val result = transformerValue.apply(Iterator(imf)).toArray.head
+
+      if (!result.contains(ImageFeature.imageTensor)) {
+        MatToTensor[Float]().transform(result)
+      }
+      DLImageSchema.imf2Row(result)
+    }
+
+    val transformUDF = SqlAdapter.getUDF(transFunc, DLImageSchema.floatSchema)
+
+    dataFrame.withColumn($(outputCol), transformUDF(dataFrame($(inputCol))))
```

Comment:
1. Use the transform in Transformer or FeatuerTransformer instead, as it will handle broadcast, clone, etc.; someting like:
   ```
   r1 = dataFrame.rdd().map(DLImageSchema.row2IMF(_)))
   r2 = localTransformer(r1).map(DLImageSchema.imf2Row(_))
   ```

2. Do we want to replicate the image contents in both inputCol and outputCol here?

Reply:
I'm thinking we should support the case the there're multiple columns in the input DataFrame. To use the RDD interface, we need to extract, transform then zip back with the other columns, which can bring extra performance cost and instability during runtime.

For 2, user may set inputCol and outputCol to the same name. I'll add a unit test.

Reply:
The current usage of transformer in the code is incorrect; you also need to clone the transformer after broadcast; see https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/Transformer.scala#L67

Reply:
Thanks for the reminder. I'll update and run some tests.